pkg/auth/ldaputil/query.go

```diff
@@ -0,0 +1,211 @@
+package ldaputil
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-ldap/ldap"
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/util"
+
+	"github.com/openshift/origin/pkg/cmd/server/api"
+)
+
+// LDAPQuery encodes an LDAP query
+type LDAPQuery struct {
+	// The DN of the branch of the directory where all searches should start from
+	BaseDN string
+
+	// The (optional) scope of the search. Defaults to the entire subtree if not set
+	Scope Scope
+
+	// The (optional) behavior of the search with regards to alisases. Defaults to always
+	// dereferencing if not set
+	DerefAliases DerefAliases
+
+	// TimeLimit holds the limit of time in seconds that any request to the server can remain outstanding
+	// before the wait for a response is given up. If this is 0, no client-side limit is imposed
+	TimeLimit int
+
+	// Filter is a valid LDAP search filter that retrieves all relevant entries from the LDAP server with the base DN
+	Filter string
+}
+
+// NewSearchRequest creates a new search request for the LDAP query and optionally includes more attributes
+func (q *LDAPQuery) NewSearchRequest(additionalAttributes []string) *ldap.SearchRequest {
+	return ldap.NewSearchRequest(
+		q.BaseDN,
+		int(q.Scope),
+		int(q.DerefAliases),
+		0, // allowed return size - indicates no limit
+		q.TimeLimit,
+		false, // not types only
+		q.Filter,
+		additionalAttributes,
+		nil, // no controls
+	)
+}
+
+// LDAPQueryOnAttribute encodes an LDAP query that conjoins two filters to extract a specific LDAP entry
+// This query is not self-sufficient and needs the value of the QueryAttribute to construct the final filter
+type LDAPQueryOnAttribute struct {
+	// Query retrieves entries from an LDAP server
+	LDAPQuery
+
+	// QueryAttribute is the attribute for a specific filter that, when conjoined with the common filter,
+	// retrieves the specific LDAP entry from the LDAP server. (e.g. "cn", when formatted with "aGroupName"
+	// and conjoined with "objectClass=groupOfNames", becomes (&(objectClass=groupOfNames)(cn=aGroupName))")
+	QueryAttribute string
+}
+
+// IdentifiyingLDAPQueryOptions holds a query and the attribute that identifies the entries that the query returns
+type IdentifiyingLDAPQueryOptions struct {
+	// Query retrieves entries from an LDAP server
+	LDAPQueryOnAttribute
+
+	// NameAttributes defines the attributes for the LDAP entries returned by the Query that will be interpreted
+	// as their names
+	NameAttributes []string
```

Comment:
If we do this, there's no need for this type, everything can use `LDAPQueryOnAttribute` and hold the `NameAttributes` separately. I like this because it nicely bundles up almost everything you need for creating search requests, but I don't have strong feelings about it.